### PR TITLE
fix multi controller for sync resource

### LINKED ISSUE #801: Sub cluster initiated multiple controllers for processing the same resource
#### What happened:
When the clusternet agent is started, the controller will be started according to the delivered resource type to synchronize the resource status, but  we find that the sub cluster occasionally initiated multiple controllers for processing the same resource .
企业微信截图_f79fb96f-0a55-4ef7-97ca-f30131019622
企业微信截图_007c1526-d98b-410a-b2eb-9ef58dc9989f

https://github.com/clusternet/clusternet/blob/main/pkg/agent/deployer/generic/generic.go#L178-L182

#### What you expected to happen:
The agent should only start one controller to handle one resource

#### Environment:
- Clusternet version:
    - Clusternet-agent version (user `clusternet-agent --version=json`):
    - Clusternet-hub version (user `clusternet-hub --version=json`):
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Others:

### PR DESCRIPTION
**Pull request by zhenkuang**
#### What type of PR is this?
Bug fixed
#### Which issue(s) this PR fixes:

Fixes #801

### CODE REVIEW

File: pkg/agent/deployer/generic/generic.go [hunk 1]
```diff
@@ -233,13 +233,14 @@ func (deployer *Deployer) ControllerHasStarted(gvk schema.GroupVersionKind) bool
 	}
 }
 
-func (deployer *Deployer) AddController(gvk schema.GroupVersionKind, controller *resourcecontroller.Controller) {
+func (deployer *Deployer) AddController(gvk schema.GroupVersionKind, controller *resourcecontroller.Controller) bool {
 	deployer.lock.Lock()
 	defer deployer.lock.Unlock()
 	if _, ok := deployer.rsControllers[gvk]; ok {
-		return
+		return false
```
Comment (dixudx) [inline pkg/agent/deployer/generic/generic.go:240]: `False` is misleading. For a single GVK, we have no idea whether we've successfully added a matching controller.

**Review by dixudx — COMMENTED** [review]
Good catch. Please optimize the logic. We're getting close.

File: pkg/agent/deployer/generic/generic.go (hunk 1 above)
Comment (zhenkuang) [reply pkg/agent/deployer/generic/generic.go:240]: If false is returned, it can be determined that a resource has triggered to start matching controller

Comment (zhenkuang) [reply pkg/agent/deployer/generic/generic.go:240]: The main issue here is that the judgment of 152 line above and the startup of controller here are not atomic operations

**Review by dixudx — APPROVED** [review]
Good work.

**Pull request merged**